[TEST-FAILURE] ReplicatedMapTest.testEvictionObject

```
java.lang.AssertionError: CountDownLatch failed to complete within 120 seconds , count left: 1
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:748)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:741)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:733)
	at com.hazelcast.replicatedmap.ReplicatedMapTest.testEviction(ReplicatedMapTest.java:967)
	at com.hazelcast.replicatedmap.ReplicatedMapTest.testEvictionObject(ReplicatedMapTest.java:934)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-IbmJDK1.7/com.hazelcast$hazelcast/742/testReport/junit/com.hazelcast.replicatedmap/ReplicatedMapTest/testEvictionObject/

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/ReplicatedMapEvictionProcessor.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/ReplicatedMapEvictionProcessor.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.replicatedmap.impl;
 
+import com.hazelcast.replicatedmap.impl.operation.EvictionOperation;
 import com.hazelcast.replicatedmap.impl.record.ReplicatedRecordStore;
 import com.hazelcast.spi.NodeEngine;
 import com.hazelcast.util.scheduler.EntryTaskScheduler;
@@ -28,30 +29,19 @@ import java.util.Collection;
  */
 public class ReplicatedMapEvictionProcessor implements ScheduledEntryProcessor<Object, Object> {
 
-    final NodeEngine nodeEngine;
-    final ReplicatedMapService replicatedMapService;
-    final String mapName;
+    private ReplicatedRecordStore store;
+    private NodeEngine nodeEngine;
+    private int partitionId;
 
-    public ReplicatedMapEvictionProcessor(NodeEngine nodeEngine, ReplicatedMapService replicatedMapService, String mapName) {
+    public ReplicatedMapEvictionProcessor(ReplicatedRecordStore store, NodeEngine nodeEngine, int partitionId) {
+        this.store = store;
         this.nodeEngine = nodeEngine;
-        this.replicatedMapService = replicatedMapService;
-        this.mapName = mapName;
+        this.partitionId = partitionId;
     }
 
     public void process(EntryTaskScheduler<Object, Object> scheduler, Collection<ScheduledEntry<Object, Object>> entries) {
-        for (ScheduledEntry<Object, Object> entry : entries) {
-            Object key = entry.getKey();
-            int partitionId = replicatedMapService.getNodeEngine().getPartitionService().getPartitionId(key);
-            final ReplicatedRecordStore store = replicatedMapService.getReplicatedRecordStore(mapName, false, partitionId);
-            if (store == null) {
-                continue;
-            }
-            if (entry.getValue() == null) {
-                store.removeTombstone(key);
-            } else {
-                store.evict(key);
-            }
-        }
+        EvictionOperation evictionOperation = new EvictionOperation(store, entries);
+        evictionOperation.setPartitionId(partitionId);
+        nodeEngine.getOperationService().executeOperation(evictionOperation);
     }
-
 }

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/operation/EvictionOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/operation/EvictionOperation.java
@@ -1,0 +1,54 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.replicatedmap.impl.operation;
+
+import com.hazelcast.replicatedmap.impl.record.ReplicatedRecordStore;
+import com.hazelcast.spi.AbstractOperation;
+import com.hazelcast.util.scheduler.ScheduledEntry;
+import java.util.Collection;
+
+/**
+ * Evicts set of entries from the record store. Runs locally.
+ */
+public class EvictionOperation extends AbstractOperation {
+
+    private final ReplicatedRecordStore store;
+    private final Collection<ScheduledEntry<Object, Object>> entries;
+
+    public EvictionOperation(ReplicatedRecordStore store, Collection<ScheduledEntry<Object, Object>> entries) {
+        this.store = store;
+        this.entries = entries;
+    }
+
+    @Override
+    public void run() throws Exception {
+        for (ScheduledEntry<Object, Object> entry : entries) {
+            Object key = entry.getKey();
+            store.evict(key);
+        }
+    }
+
+    @Override
+    public boolean returnsResponse() {
+        return false;
+    }
+
+    @Override
+    public boolean validatesTarget() {
+        return false;
+    }
+}

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/operation/SyncReplicatedMapDataOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/operation/SyncReplicatedMapDataOperation.java
@@ -67,6 +67,9 @@ public class SyncReplicatedMapDataOperation<K, V> extends AbstractOperation {
             if (oldRecord != null) {
                 replicatedRecord.setHits(oldRecord.getHits());
             }
+            if (record.getTtl() > 0) {
+                store.scheduleTtlEntry(record.getTtl(), key, value);
+            }
             newStorage.putInternal(key, replicatedRecord);
         }
         newStorage.setVersion(version);

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractBaseReplicatedRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractBaseReplicatedRecordStore.java
@@ -39,7 +39,7 @@ import java.util.concurrent.atomic.AtomicReference;
  * @param <K> key type
  * @param <V> value type
  */
-abstract class AbstractBaseReplicatedRecordStore<K, V> implements ReplicatedRecordStore {
+public abstract class AbstractBaseReplicatedRecordStore<K, V> implements ReplicatedRecordStore {
 
     protected final AtomicReference<InternalReplicatedMapStorage<K, V>> storageRef;
     protected final ReplicatedMapService replicatedMapService;
@@ -67,8 +67,7 @@ abstract class AbstractBaseReplicatedRecordStore<K, V> implements ReplicatedReco
         this.storageRef.set(new InternalReplicatedMapStorage<K, V>());
         this.ttlEvictionScheduler = EntryTaskSchedulerFactory
                 .newScheduler(nodeEngine.getExecutionService().getDefaultScheduledExecutor(),
-                        new ReplicatedMapEvictionProcessor(nodeEngine, replicatedMapService, name)
-                        , ScheduleType.POSTPONE);
+                        new ReplicatedMapEvictionProcessor(this, nodeEngine, partitionId), ScheduleType.POSTPONE);
     }
 
     public InternalReplicatedMapStorage<K, V> getStorage() {
@@ -107,12 +106,14 @@ abstract class AbstractBaseReplicatedRecordStore<K, V> implements ReplicatedReco
     }
 
 
-    ScheduledEntry<K, V> cancelTtlEntry(K key) {
+    @Override
+    public ScheduledEntry cancelTtlEntry(Object key) {
         return ttlEvictionScheduler.cancel(key);
     }
 
-    boolean scheduleTtlEntry(long delayMillis, K key, V object) {
-        return ttlEvictionScheduler.schedule(delayMillis, key, object);
+    @Override
+    public boolean scheduleTtlEntry(long delayMillis, Object key, Object value) {
+        return ttlEvictionScheduler.schedule(delayMillis, key, value);
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractReplicatedRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractReplicatedRecordStore.java
@@ -55,17 +55,6 @@ public abstract class AbstractReplicatedRecordStore<K, V> extends AbstractBaseRe
     }
 
     @Override
-    public void removeTombstone(Object key) {
-        isNotNull(key, "key");
-        K marshalledKey = (K) marshall(key);
-        ReplicatedRecord<K, V> current = getStorage().get(marshalledKey);
-        if (current == null || current.getValueInternal() != null) {
-            return;
-        }
-        getStorage().remove(marshalledKey, current);
-    }
-
-    @Override
     public Object remove(Object key) {
         isNotNull(key, "key");
         long time = Clock.currentTimeMillis();
@@ -146,19 +135,19 @@ public abstract class AbstractReplicatedRecordStore<K, V> extends AbstractBaseRe
         K marshalledKey = (K) marshall(key);
         V marshalledValue = (V) marshall(value);
         final long ttlMillis = ttl == 0 ? 0 : timeUnit.toMillis(ttl);
-        final ReplicatedRecord old = getStorage().get(marshalledKey);
+        final ReplicatedRecord<K, V> old = getStorage().get(marshalledKey);
         ReplicatedRecord<K, V> record = old;
         if (old == null) {
             record = buildReplicatedRecord(marshalledKey, marshalledValue, ttlMillis);
             getStorage().put(marshalledKey, record);
         } else {
-            oldValue = (V) old.getValueInternal();
+            oldValue = old.getValueInternal();
             if (incrementHits) {
-                getStorage().get(marshalledKey).setValue(marshalledValue, ttlMillis);
+                old.setValue(marshalledValue, ttlMillis);
             } else {
-                getStorage().get(marshalledKey).setValueInternal(marshalledValue, ttlMillis);
+                old.setValueInternal(marshalledValue, ttlMillis);
             }
-            getStorage().incrementVersion();
+            getStorage().put(marshalledKey, old);
         }
         if (ttlMillis > 0) {
             scheduleTtlEntry(ttlMillis, marshalledKey, marshalledValue);

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/ReplicatedRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/ReplicatedRecordStore.java
@@ -17,6 +17,7 @@
 package com.hazelcast.replicatedmap.impl.record;
 
 import com.hazelcast.replicatedmap.merge.ReplicatedMapMergePolicy;
+import com.hazelcast.util.scheduler.ScheduledEntry;
 import java.util.Collection;
 import java.util.Comparator;
 import java.util.Iterator;
@@ -33,8 +34,6 @@ public interface ReplicatedRecordStore {
     Object remove(Object key);
 
     void evict(Object key);
-
-    void removeTombstone(Object key);
 
     Object get(Object key);
 
@@ -79,6 +78,10 @@ public interface ReplicatedRecordStore {
     void putRecord(RecordMigrationInfo record);
 
     InternalReplicatedMapStorage getStorage();
+
+    ScheduledEntry<Object, Object> cancelTtlEntry(Object key);
+
+    boolean scheduleTtlEntry(long delayMillis, Object key, Object object);
 
     boolean isLoaded();
 

--- a/hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapBaseTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapBaseTest.java
@@ -3,6 +3,7 @@ package com.hazelcast.replicatedmap;
 import com.hazelcast.config.Config;
 import com.hazelcast.config.InMemoryFormat;
 import com.hazelcast.config.ReplicatedMapConfig;
+import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.ReplicatedMap;
 import com.hazelcast.replicatedmap.impl.ReplicatedMapProxy;
 import com.hazelcast.replicatedmap.impl.ReplicatedMapService;
@@ -11,6 +12,8 @@ import com.hazelcast.replicatedmap.impl.record.ReplicatedRecordStore;
 import com.hazelcast.test.HazelcastTestSupport;
 import java.lang.reflect.Field;
 import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Random;
 
 import static org.junit.Assert.fail;
@@ -72,6 +75,14 @@ public abstract class ReplicatedMapBaseTest extends HazelcastTestSupport {
     }
 
     @SuppressWarnings("unchecked")
+    protected <K, V> ReplicatedRecordStore getStore(ReplicatedMap<K, V> map, K key) throws Exception {
+        ReplicatedMapProxy<K, V> proxy = (ReplicatedMapProxy<K, V>) map;
+        ReplicatedMapService service = (ReplicatedMapService) REPLICATED_MAP_SERVICE.get(proxy);
+        return service.getReplicatedRecordStore(map.getName(), false, key);
+    }
+
+
+    @SuppressWarnings("unchecked")
     protected AbstractMap.SimpleEntry<Integer, Integer>[] buildTestValues() {
         Random random = new Random();
         AbstractMap.SimpleEntry<Integer, Integer>[] testValues = new AbstractMap.SimpleEntry[100];
@@ -79,5 +90,23 @@ public abstract class ReplicatedMapBaseTest extends HazelcastTestSupport {
             testValues[i] = new AbstractMap.SimpleEntry<Integer, Integer>(random.nextInt(), random.nextInt());
         }
         return testValues;
+    }
+
+    public List<ReplicatedMap> createMapOnEachInstance(HazelcastInstance[] instances, String replicatedMapName) {
+        ArrayList<ReplicatedMap> maps = new ArrayList<ReplicatedMap>();
+        for (int i = 0; i < instances.length; i++) {
+            ReplicatedMap<Object, Object> replicatedMap = instances[i].getReplicatedMap(replicatedMapName);
+            maps.add(replicatedMap);
+        }
+        return maps;
+    }
+
+    public ArrayList<Integer> generateRandomIntegerList(int count) {
+        final ArrayList<Integer> keys = new ArrayList<Integer>();
+        final Random random = new Random();
+        for (int i = 0; i < count; i++) {
+            keys.add(random.nextInt());
+        }
+        return keys;
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapTtlTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapTtlTest.java
@@ -1,0 +1,73 @@
+package com.hazelcast.replicatedmap;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.ReplicatedMap;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.SlowTest;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category(value = {SlowTest.class, ParallelTest.class})
+public class ReplicatedMapTtlTest extends ReplicatedMapBaseTest {
+
+    @Test
+    public void testPutWithTTL() throws Exception {
+        int nodeCount = 5;
+        int keyCount = 20000;
+        int operationCount = 20000;
+        int threadCount = 15;
+        int ttl = 500;
+        TimeUnit timeUnit = TimeUnit.MILLISECONDS;
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory();
+        HazelcastInstance[] instances = factory.newInstances(null, nodeCount);
+        String mapName = randomMapName();
+        List<ReplicatedMap> maps = createMapOnEachInstance(instances, mapName);
+        ArrayList<Integer> keys = generateRandomIntegerList(keyCount);
+        Thread[] threads = createThreads(threadCount, maps, keys, ttl, timeUnit, operationCount);
+        for (Thread thread : threads) {
+            thread.start();
+        }
+        for (Thread thread : threads) {
+            thread.join();
+        }
+        for (ReplicatedMap map : maps) {
+            assertSizeEventually(0, map, 60);
+        }
+
+    }
+
+    private Thread[] createThreads(int count, List<ReplicatedMap> maps, ArrayList<Integer> keys,
+                                   long ttl, TimeUnit timeunit, int operations) {
+        Thread[] threads = new Thread[count];
+        int size = maps.size();
+        for (int i = 0; i < count; i++) {
+            threads[i] = createPutOperationThread(maps.get(i % size), keys, ttl, timeunit, operations);
+        }
+        return threads;
+    }
+
+    private Thread createPutOperationThread(final ReplicatedMap<String, Object> map, final ArrayList<Integer> keys,
+                                            final long ttl, final TimeUnit timeunit, final int operations) {
+        return new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Random random = new Random();
+                int size = keys.size();
+                for (int i = 0; i < operations; i++) {
+                    int index = i % size;
+                    String key = "foo-" + keys.get(index);
+                    map.put(key, random.nextLong(), 1 + random.nextInt((int) ttl), timeunit);
+                }
+            }
+        });
+    }
+
+}

--- a/hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapWriteOrderTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapWriteOrderTest.java
@@ -20,14 +20,8 @@ import com.hazelcast.config.Config;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.ReplicatedMap;
 import com.hazelcast.instance.HazelcastInstanceFactory;
-import com.hazelcast.replicatedmap.impl.ReplicatedMapProxy;
-import com.hazelcast.replicatedmap.impl.ReplicatedMapService;
-import com.hazelcast.replicatedmap.impl.record.ReplicatedRecord;
-import com.hazelcast.replicatedmap.impl.record.ReplicatedRecordStore;
 import com.hazelcast.test.AssertTask;
-import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
-import java.lang.reflect.Field;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -42,19 +36,7 @@ import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
 
 @RunWith(Parameterized.class)
-public class ReplicatedMapWriteOrderTest extends HazelcastTestSupport {
-
-    protected static Field REPLICATED_MAP_SERVICE;
-
-    static {
-        try {
-            REPLICATED_MAP_SERVICE = ReplicatedMapProxy.class.getDeclaredField("service");
-            REPLICATED_MAP_SERVICE.setAccessible(true);
-        } catch (Exception e) {
-            throw new RuntimeException(e);
-        }
-    }
-
+public class ReplicatedMapWriteOrderTest extends ReplicatedMapBaseTest {
 
     int nodeCount;
     int operations;
@@ -102,7 +84,7 @@ public class ReplicatedMapWriteOrderTest extends HazelcastTestSupport {
         TestHazelcastInstanceFactory factory = new TestHazelcastInstanceFactory(nodeCount);
         final HazelcastInstance[] instances = factory.newInstances(config);
         String replicatedMapName = "test";
-        final List<ReplicatedMap> maps = createMaps(instances, replicatedMapName);
+        final List<ReplicatedMap> maps = createMapOnEachInstance(instances, replicatedMapName);
         ArrayList<Integer> keys = generateRandomIntegerList(keyCount);
         Thread[] threads = createThreads(nodeCount, maps, keys, operations);
         for (Thread thread : threads) {
@@ -153,24 +135,6 @@ public class ReplicatedMapWriteOrderTest extends HazelcastTestSupport {
         return threads;
     }
 
-    private List<ReplicatedMap> createMaps(HazelcastInstance[] instances, String replicatedMapName) {
-        ArrayList<ReplicatedMap> maps = new ArrayList<ReplicatedMap>();
-        for (int i = 0; i < instances.length; i++) {
-            ReplicatedMap<Object, Object> replicatedMap = instances[i].getReplicatedMap(replicatedMapName);
-            maps.add(replicatedMap);
-        }
-        return maps;
-    }
-
-    private ArrayList<Integer> generateRandomIntegerList(int count) {
-        final ArrayList<Integer> keys = new ArrayList<Integer>();
-        final Random random = new Random();
-        for (int i = 0; i < count; i++) {
-            keys.add(random.nextInt());
-        }
-        return keys;
-    }
-
     private Thread createPutOperationThread(final ReplicatedMap<String, Object> map, final ArrayList<Integer> keys,
                                             final int operations) {
         return new Thread(new Runnable() {
@@ -188,21 +152,5 @@ public class ReplicatedMapWriteOrderTest extends HazelcastTestSupport {
             }
         });
     }
-
-    @SuppressWarnings("unchecked")
-    protected <K, V> ReplicatedRecord<K, V> getReplicatedRecord(ReplicatedMap<K, V> map, K key) throws Exception {
-        ReplicatedMapProxy<K, V> proxy = (ReplicatedMapProxy<K, V>) map;
-        ReplicatedMapService service = (ReplicatedMapService) REPLICATED_MAP_SERVICE.get(proxy);
-        ReplicatedRecordStore store = service.getReplicatedRecordStore(map.getName(), false, key);
-        return store.getReplicatedRecord(key);
-    }
-
-    @SuppressWarnings("unchecked")
-    protected <K, V> ReplicatedRecordStore getStore(ReplicatedMap<K, V> map, K key) throws Exception {
-        ReplicatedMapProxy<K, V> proxy = (ReplicatedMapProxy<K, V>) map;
-        ReplicatedMapService service = (ReplicatedMapService) REPLICATED_MAP_SERVICE.get(proxy);
-        return service.getReplicatedRecordStore(map.getName(), false, key);
-    }
-
 
 }

--- a/hazelcast/src/test/java/com/hazelcast/replicatedmap/impl/record/LazyIteratorTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/replicatedmap/impl/record/LazyIteratorTest.java
@@ -21,6 +21,7 @@ import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.util.scheduler.ScheduledEntry;
 import java.util.Collection;
 import java.util.Comparator;
 import java.util.HashSet;
@@ -565,11 +566,6 @@ public class LazyIteratorTest
         }
 
         @Override
-        public void removeTombstone(Object key) {
-
-        }
-
-        @Override
         public Object get(Object key) {
             return null;
         }
@@ -675,6 +671,16 @@ public class LazyIteratorTest
         @Override
         public InternalReplicatedMapStorage getStorage() {
             return null;
+        }
+
+        @Override
+        public ScheduledEntry<Object, Object> cancelTtlEntry(Object key) {
+            return null;
+        }
+
+        @Override
+        public boolean scheduleTtlEntry(long delayMillis, Object key, Object object) {
+            return false;
         }
 
         @Override

